internal/resources/syntheticmonitoring/resource_probe.go

```diff
@@ -163,14 +163,31 @@ func ResourceProbeUpdate(ctx context.Context, d *schema.ResourceData, meta inter
 
 func ResourceProbeDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	c := meta.(*common.Client).SMAPI
-	var diags diag.Diagnostics
 	id, _ := strconv.ParseInt(d.Id(), 10, 64)
-	err := c.DeleteProbe(ctx, id)
+
+	// Remove the probe from any checks that use it.
+	checks, err := c.ListChecks(ctx)
 	if err != nil {
 		return diag.FromErr(err)
 	}
+	for _, check := range checks {
+		for i, probeID := range check.Probes {
+			if probeID == id {
+				check.Probes = append(check.Probes[:i], check.Probes[i+1:]...)
+				if len(check.Probes) == 0 {
+					return diag.Errorf(`could not delete probe %d. It is the only probe for check %q.
```

Comment:
I understand that when a probe has to be recreated, the way Terraform works does not allow to create the new probe first and then move the checks to the new probe before deleting the previous one?

Reply:
The problem stems from the following:
- When "tainting" a resource, Terraform can either create it first and then destroy the old one (opt-in) or the opposite (default behavior).
- The "create" function does not have any knowledge that a "delete" came beforehand, or will come afterwards
- Creating the probe first would mean creating under a temporary name to avoid conflicts. This is generally bad practice because we're dealing with as-code here and any interruption in the process would leave unmanaged things with temporary names. Not to mention that it'd be impossible for the create function to actually know it has to use a temporary name (because of the previous point)

Destroying the probe first is the default behaviour and is the one I'm trying to fix here. IMO, it's alright to have the limitation that every check needs to have at least one probe assigned, which is why I'm closing the issue. We just need to do the best effort, and document (with errors) what the best way to go forward is for the user, if it does still fail.

Reply:
Got it. Thanks for the explanation.